go-controller/pkg/node/port_claim.go

```diff
@@ -0,0 +1,181 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+var port localPort
+
+type activeSocket interface {
+	Close() error
+}
+
+type localPort interface {
+	open(port int32, protocol kapi.Protocol, svc *kapi.Service) error
+	close(port int32, svc *kapi.Service) error
+}
+
+type portClaimWatcher struct {
+	recorder      record.EventRecorder
+	activeSockets map[int32]activeSocket
+}
+
+func newPortClaimWatcher(recorder record.EventRecorder) localPort {
+	return &portClaimWatcher{
+		recorder:      recorder,
+		activeSockets: make(map[int32]activeSocket),
+	}
+}
+
+func initPortClaimWatcher(recorder record.EventRecorder, wf *factory.WatchFactory) error {
+	port = newPortClaimWatcher(recorder)
+	_, err := wf.AddServiceHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			if errors := addServicePortClaim(svc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldSvc := old.(*kapi.Service)
+			newSvc := new.(*kapi.Service)
+			if errors := updateServicePortClaim(oldSvc, newSvc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			if errors := deleteServicePortClaim(svc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+	}, nil)
+	return err
+}
+
+func addServicePortClaim(svc *kapi.Service) []error {
+	return handleServicePortClaim(svc, true)
```

Comment:
i don't really like handle functions that handle CRUD operations. I would rather it be split up into each method. The service code got really crazy over time trying to understand the munged paths of subsequent handlers getting called during any CRUD operation.